src/components/ValueAnnotationTabs.tsx

```diff
@@ -0,0 +1,101 @@
+  onUpdateLevelTerm: (
+    columnID: string,
+    value: string,
+    term: { identifier: string; label: string } | null
+  ) => void;
+}
+
+function ValueAnnotationTabs({
+  columns,
+  dataTable,
+  onUpdateDescription,
+  onUpdateUnits,
+  onToggleMissingValue,
+  onUpdateFormat,
+  onUpdateLevelTerm,
+}: ValueAnnotationTabsProps) {
+  const [activeTab, setActiveTab] = useState(0);
+  const columnEntries = Object.entries(columns);
+  const columnIds = Object.keys(columns);
+
+  const handleTabChange = (_: React.SyntheticEvent, newValue: number) => {
+    setActiveTab(newValue);
+  };
+
+  if (columnIds.length === 0) {
+    return <Paper elevation={3}>No columns to display</Paper>;
+  }
+
+  return (
+    <Paper elevation={3} className="h-full flex flex-col">
+      <Tabs value={activeTab} onChange={handleTabChange} variant="scrollable" scrollButtons="auto">
+        {columnIds.map((columnId) => (
+          <Tab
+            key={columnId}
+            label={columns[columnId].header || columnId}
+            data-cy={`${columnId}-tab`}
+          />
+        ))}
+      </Tabs>
+      <div className="flex-1 overflow-auto p-2">
+        {columnEntries.map(([columnId, column], index) => {
+          const uniqueValues = dataTable[columnId] ? Array.from(new Set(dataTable[columnId])) : [];
+
+          return (
+            <div
+              key={columnId}
+              role="tabpanel"
+              hidden={activeTab !== index}
+              id={`tabpanel-${columnId}`}
+              aria-labelledby={`tab-${columnId}`}
+            >
+              {activeTab === index && (
+                <div className="h-full">
+                  {column.dataType === 'Categorical' ? (
+                    <Categorical
+                      columnID={columnId}
+                      uniqueValues={uniqueValues}
+                      levels={column.levels || {}}
+                      standardizedVariable={column.standardizedVariable}
+                      missingValues={column.missingValues || []}
+                      onUpdateDescription={onUpdateDescription}
+                      onToggleMissingValue={onToggleMissingValue}
+                      onUpdateLevelTerm={onUpdateLevelTerm}
+                    />
+                  ) : (
+                    <Continuous
+                      columnID={columnId}
+                      units={column.units || ''}
+                      missingValues={column.missingValues || []}
+                      columnValues={dataTable[columnId] || []}
+                      format={column.format}
+                      standardizedVariable={column.standardizedVariable}
+                      onUpdateUnits={onUpdateUnits}
+                      onToggleMissingValue={onToggleMissingValue}
+                      onUpdateFormat={onUpdateFormat}
+                    />
+                  )}
+                </div>
+              )}
+            </div>
+          );
+        })}
```

Comment:
Why is this thing not a component? I'm really not a big fan of this unneccesary complexity stuff with large inline components and so on. Please make sure you keep things simple first, and then worry about edge cases and prettyness and so on

Reply:
I'm not sure what you mean here. This entire component was initially part of the valueAnnotation and was factored out into its own component. If we were to factor out the logic you highlighted the parent component would be quite empty.